text/0024-schema-syntax.md

```diff
@@ -0,0 +1,449 @@
+    action ModifyGroup, DeleteGroup 
+        appliesTo { principal: [User], resource: [Group] };
+}
+```
+
+### Mockup: GitHub Model
+
+Here is a version of the [GitHub](https://github.com/cedar-policy/cedar-examples/blob/main/cedar-example-use-cases/github_example/schema.json) schema.
+
+```
+namespace GitHub {
+    entity User in [UserGroup,Team];
+    entity UserGroup in [UserGroup];
+    entity Repository {
+        readers: UserGroup,
+        triagers: UserGroup,
+        writers: UserGroup,
+        maintainers: UserGroup,
+        admins: UserGroup   
+    };
+    entity Issue {
+        repo: Repository,
+        reporter: User
+    };
+    entity Org {
+        members: UserGroup,
+        owners: UserGroup,
+        memberOfTypes: UserGroup
+    };
+    action pull, fork, push, add_reader, add_triager, add_writer, add_maintainer, add_admin 
+        appliesTo { principal: [User], resource: [Repository] };
+    action delete_issue, edit_issue, assign_issue 
+        appliesTo { principal: [User], resource: [Issue] };
+}
+```
+
+### Grammar
+
+Here is a grammar for the proposed schema syntax.
+
+The grammar applies the following the conventions. Capitalized words stand for grammar productions, and lexical tokens are given in all-caps. When productions or tokens match those in the Cedar policy grammar, we use the same names (e.g., `IDENT` and `Path`).
+
+For grammar productions it uses `|` for alternatives, `[]` for optional productions, `()` for grouping, and `{}` for repetition of a form zero or more times.
+
+Tokens are defined using regular expressions, where `[]` stands for character ranges; `|` stands for alternation; `*` , `+` , and `?` stand for zero or more, one or more, and zero or one occurrences, respectively; `~` stands for complement; and `-` stands for difference. The grammar ignores whitespace and comments.
+
+```
+Schema    := {Namespace}
+Namespace := ('namespace' Path '{' {Decl} '}') | {Decl}
+Decl      := Entity | Action | TypeDecl
+Entity    := 'entity' Idents ['in' EntOrTyps] [['='] RecType] ';'
+Action    := 'action' Names ['in' (Name | '[' [Names] ']')] [AppliesTo] [ActAttrs]';'
+TypeDecl  := 'type' IDENT '=' Type ';'
+Type      := PRIMTYPE | IDENT | SetType | RecType
+EntType   := Path
+SetType   := 'Set' '<' Type '>'
+RecType   := '{' [AttrDecls] '}'
+AttrDecls := Name ['?'] ':' Type [',' | ',' AttrDecls]
+AppliesTo := 'appliesTo' '{' AppDecls '}'
+ActAttrs  := 'attributes' '{' AttrDecls '}'
+AppDecls  := VAR ':' EntOrTyps [',' | ',' AppDecls]
+Path      := IDENT {'::' IDENT}
+EntTypes  := Path {',' Path}
+EntOrTyps := EntType | '[' [EntTypes] ']'
+Name      := IDENT | STR
+Names     := Name {',' Name}
+Idents    := IDENT {',' IDENT}
+
+IDENT     := ['_''a'-'z''A'-'Z']['_''a'-'z''A'-'Z''0'-'9']* - PRIMTYPE
+STR       := Fully-escaped Unicode surrounded by '"'s
+PRIMTYPE  := 'Long' | 'String' | 'Bool'
+VAR       := 'principal' | 'action' | 'resource' | 'context'
```

Comment:
It doesn't need to contain `action`, right?

Reply:
Thanks for these. Updated the grammar.